servlet/spring-boot/java/saml2/refreshable-metadata/src/main/java/example/RefreshableRelyingPartyRegistrationRepository.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package example;
+
+import java.io.InputStream;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Spliterator;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.io.Resource;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.security.saml2.core.Saml2X509Credential;
+import org.springframework.security.saml2.provider.service.registration.RelyingPartyRegistration;
+import org.springframework.security.saml2.provider.service.registration.RelyingPartyRegistrationRepository;
+import org.springframework.security.saml2.provider.service.registration.RelyingPartyRegistrations;
+import org.springframework.stereotype.Component;
+
+@Component
+public class RefreshableRelyingPartyRegistrationRepository
+		implements RelyingPartyRegistrationRepository, Iterable<RelyingPartyRegistration> {
+
+	private final Map<String, RelyingPartyRegistration> relyingPartyRegistrations = new ConcurrentHashMap<>();
+
+	private final RSAPrivateKey privateKey;
+
+	private final Resource certificateResource;
+
+	public RefreshableRelyingPartyRegistrationRepository(
+			@Value("classpath:credentials/rp-private.key") RSAPrivateKey privateKey,
+			@Value("classpath:credentials/rp-certificate.crt") Resource certificateResource) {
+		this.privateKey = privateKey;
+		this.certificateResource = certificateResource;
+		refreshMetadata();
+	}
+
+	@Override
+	public RelyingPartyRegistration findByRegistrationId(String registrationId) {
+		return this.relyingPartyRegistrations.get(registrationId);
+	}
+
+	@Override
+	public Iterator<RelyingPartyRegistration> iterator() {
+		return this.relyingPartyRegistrations.values().iterator();
+	}
+
+	@Override
+	public void forEach(Consumer<? super RelyingPartyRegistration> action) {
```

Comment:
What does implementing `forEach` and `spliterator` buy you? It seems like the default implementations do the same thing already.

Reply:
Probably an oversight. Didn't notice that there was default implementations.